### ISSUE TITLE
Auto-update does not work, because Warp claims not to have permission to update, even when it (apparently) does

### ISSUE THREAD

**Issue by shrugalic** [warpdotdev/Warp#3021]
### Describe the bug

I get the following message on every update:

> A new version of Warp is available but Warp does not have permission to update.

Warp sits in a folder that belongs to my (non-admin) user and has full permissions in there, so from my point of view it should be able to update.

I suspect it might be because I'm running as a standard non-admin user, and thus do not have write access to `/Applications`.

### To Reproduce

1. Run as a standard non-admin user
2. Run Warp from somewhere within your home folder, where your user has full access (say `~/Applications/…/`)
3. Release an update
4. Observe the behavior described above

### Expected behavior

Because Warp is located somewhere in within my home folder (`~/Applications/…/`), my expectation is that it should be able to update automatically.

### Operating System

MacOS

### Operating System and Version

Ventura 13.3.1

### Shell Version

fish, version 3.6.0

### Warp Version

v0.2023.05.02.08.03.stable_00

### Does this block you from using Warp daily?

No

### Is this a Warp specific issue? (i.e. does it happen in Terminal, iTerm, Kitty, etc.)

N/A, the issue only applies to a Warp specific feature. (i.e. Warp AI, Workflows)

### Warp Internal (ignore): linear-label:b8107fdf-ba31-488d-b103-d271c89cac3e

None

**Comment by dannyneira** [comment]
Hey @shrugalic, Could you please check this doc to see if it helps? https://docs.warp.dev/help/updating-warp#auto-update-issues

**Comment by shrugalic** [comment]
> Hey @shrugalic, Could you please check this doc to see if it helps? https://docs.warp.dev/help/updating-warp#auto-update-issues

Thanks, I'll give it a try on the next update. The relevant part is this:
> You are a non-Admin user. […] If you have admin access on the computer, opening the app with the admin user should fix the auto-update issues.

**Comment by challsealed** [comment]
What about those who are not admins?

**Comment by shrugalic** [comment]
OK, so today's update failed with the usual error message, so I switched over to an admin account, and opened Warp using `sudo open /Users/my-non-admin-user/Applications/Warp.app`.

It wanted me to sign in before doing anything, which is a pain in the ass, because the admin user does not actually need to use Warp, nor does it need to know my login. I did it anyway.

Interestingly enough, it showed the same update error message when run by an admin account!

I copied it to `/Users/Shared` and tried once more: Same error message yet again.

I then copied it to `/Applications`, and lo and behold, it did **not** show the error message, but instead displayed the orange-ish "update available" bubble in the top right of its window.

I did **not** update as admin user though, but switched back to my non-admin user, opened the copy of Warp in `/Applications`, and it immediately displayed its release notes, apparently having automatically updated without any issues.

So for some reason Warp _really_ wants to live in `/Applications`!?

I usually prefer to keep important applications in my home folder, but mostly because it normally _helps_ with auto/self-updating apps – unlike in this case, where it prevented auto-update.

# TL;DR
Workaround: Move Warp to `/Applications`, open as an admin user once, then auto-update should work for non-admin users as well – despite non-admin-users not having write access to `/Applications`.

**Comment by shrugalic** [comment]
Given today's experiences I would slightly change the issue description:

# Warp cannot self-update when it does not live within `/Applications` (at least for non-admin users)

**Comment by shrugalic** [comment]
Having experienced a few updates to Warp, I have to edit this issue yet again:

# Self-updating as a non-admin user does not work, even if Warp is installed in `/Applications`.

**Comment by bytebuddyai** [comment]
~/Applications isn't standard and doesn't really work anymore. 

Additionally, the More Info doc is not accurate and does not give a reliable workaround: https://docs.warp.dev/help/updating-warp#auto-update-issues

The main cause of this issue is the launching user not having write access to /Applications/Warp.app and its subdirs. Default posix permissions for /Applications is 755 (rwxr-x-r-x) which allows the owner to read,write,execute but _group_ and _other_ can only read and execute. It doesn't matter if you launch Warp with an "admin user" unless that specific admin user has write access to /Applications/Warp.app/

For example, the owner of /Applications/Warp.app in the snippet below is 'rdavies' which is an "admin" user, but so is 'ryan'. In this scenario, only rdavies or root will be able to update Warp, or in other words, write to /Applications/Warp.app/ and its subdirs.

`ryan@Ryans-MacBook-Pro ~ % ls -l /Applications/Warp.app
total 0
drwxr-xr-x  7 rdavies  admin  224 Aug 10 09:10 Contents
ryan@Ryans-MacBook-Pro ~ % whoami
ryan`

If you want auto update to work in this scenario, there are three options:

1. Run warp as rdavies
2. Run warp as root (which would be done using sudo to launch Warp from the terminal)
3. Change the owner on /Applications/Warp.app and its subdirs to ryan: sudo chown -R ryan /Applications/Warp.app

**Comment by shrugalic** [comment]
As suspected, automatically updating Warp in `/Applications` as a non-admin user still does not work, despite:
1. the non-admin user being the owner of the directory tree rooted at `/Applications/Warp.app`
2. _and_ the non-admin user's group having write permissions in `/Applications`. Aside: I don't see why this would strictly be necessary, given `/Applications/Warp.app` is an application bundle – basically a folder, which does not need not be changed on update; only some of its contents will.

It doesn't even work if the non-admin user is the owner of `/Applications` itself. It seems to just plain not work as non-admin user, no matter what.

**Comment by fjeldstad** [comment]
I'm also running Warp as a non-admin user and experience the same message about it being unable to update, requiring a manual update. The question is, how do other applications solve this? For apps like Spotify, Slack, Postman etc. I get a prompt requiring me to input an admin user's credentials, then it proceeds with the update. In my opinion that would be an OK user experience for Warp as well - much better than manually having to quit and replace the app every time.

**Comment by ataetgi** [comment]
i agree with @fjeldstad, manually updating is really frustrating

**Comment by shrugalic** [comment]
For some convenience I created the following `update_Warp.command` script to download and install Warp:
```sh
#!/bin/sh

DISK_IMAGE=~/Downloads/Warp_latest.dmg
INSTALLED_APP=~/Apps/Utility/Warp.app

rm -f $DISK_IMAGE
curl "https://app.warp.dev/download?package=dmg" --location --output $DISK_IMAGE && \
hdiutil attach $DISK_IMAGE && \
killall stable
rm -Rf $INSTALLED_APP && \
cp -R /Volumes/Warp/Warp.app $INSTALLED_APP && \
hdiutil eject /Volumes/Warp && \
rm -f $DISK_IMAGE && \
open $INSTALLED_APP

exit 0
```

If you are wondering, `stable` is the name of the Warp binary.

**Comment by dannyneira** [comment]
Thanks for submitting this @shrugalic 

We already have a request tracking this over at https://github.com/warpdotdev/Warp/issues/3947

If you could please give it a 👍 and subscribe to that request, it would help us gauge interest. We'll post any updates there.

We'll be closing this as a duplicate.
